modules/tinymce/src/plugins/visualchars/main/ts/core/Nodes.ts

```diff
@@ -11,17 +11,32 @@ const isMatch = (n: SugarElement<Node>): n is SugarElement<Text> => {
     Data.regExp.test(value);
 };
 
-// inlined sugars PredicateFilter.descendants for file size
-const filterDescendants = <T extends Node>(scope: SugarElement<Node>, predicate: (x: SugarElement<Node>) => x is SugarElement<T>): SugarElement<T>[] => {
+const isContentEditableFalse = (node: SugarElement<Node>) => SugarNode.isHTMLElement(node) && ContentEditable.getRaw(node) === 'false';
+
+const isChildEditable = (node: SugarElement<Node>, currentState: boolean) => {
+  if (SugarNode.isHTMLElement(node)) {
+    const value = ContentEditable.getRaw(node);
+    if (value === 'true') {
+      return true;
+    } else if (value === 'false') {
+      return false;
+    }
+  }
+
+  return currentState;
+};
+
+// inlined sugars PredicateFilter.descendants for file size but also make it only act on editable nodes
+const filterDescendants = <T extends Node>(scope: SugarElement<Node>, predicate: (x: SugarElement<Node>) => x is SugarElement<T>, editable: boolean): SugarElement<T>[] => {
```

Comment:
It took me a while to process this, so a comment might be useful for the future. Is the basic idea 

* it's just normal PredicateFilter.descendants except 
** it skips over including any elements that are CEF, but doesn't automatically skip their children
** it skips over their children only if they don't have a contenteditable=true

So we can't just filter it afterwards, because we need to know the ancestry of each one, so instead we just manually skip anything that has not overridden a CEF ancestor by passing through the current "editable" state. Is that right?

Reply:
I thought the test for this function and the `only act on editable nodes` would be enough. Yeah it checks for editable nodes by tracking the contenteditable state if it changes between true/false if it doesn't then it uses what ever state that is on the stack. It's only when it goes down or up the state change matters down it's done by reading what that node we are going down by has. Up it just pops the stack.

I renamed the function to `filterEditableDescendants` instead so that it more clear what the function is doing and I also added more details to the comment.